### ISSUE TITLE
Fetcher cache can end up marking mutable urls as immutable

### ISSUE THREAD

**Issue by infinisil** [NixOS/nix#9905]
**Describe the bug**

I have no idea how this happened, but apparently there's an immutable entry to a master tarball in my fetcher cache:

```
$ sqlite3 ~/.cache/nix/fetcher-cache-v1.sqlite "select * from cache where input like '%master%' and immutable = 1"
-- Loading resources from /home/tweagysil/.sqliterc
input                                                         info                                                          path                                                immutable  timestamp 
------------------------------------------------------------  ------------------------------------------------------------  --------------------------------------------------  ---------  ----------
{"name":"source","type":"file","url":"https://github.com/edo  {"etag":"\"6704e1eaab5bc3c19f56a5264103e3b312b09c0b517112a4a  /nix/store/maj1sjdrrgdp1m776fjwmynr6x74j6an-source  1          1679943235
lstra/flake-compat/archive/master.tar.gz"}                    d2365fbb1db19b3\"","url":"https://codeload.github.com/edolst                                                                           
                                                              ra/flake-compat/tar.gz/refs/heads/master"}                                                                                             

{"name":"source","type":"tarball","url":"https://github.com/  {"etag":"\"6704e1eaab5bc3c19f56a5264103e3b312b09c0b517112a4a  /nix/store/r2ip1850igy8kciyaagw502s3c6ph1s4-source  1          1679943235
edolstra/flake-compat/archive/master.tar.gz"}                 d2365fbb1db19b3\"","lastModified":1673956053}
```

I have no idea how this ended up in there, and I can't reproduce it (at least with the current Nix version) if I just delete that cache entry.

Notably it happens with `flake-compat`, which has traditionally been fetched like
```nix
(import (fetchTarball "https://github.com/edolstra/flake-compat/archive/master.tar.gz") {
  src = ./.;
}).defaultNix
```

Which means that many of these instances were never pinned. This is likely related.

I doubt that it's a random unprompted bit flip, because it's 2 (effectively the same) cache entries with the same problem.

**Expected behavior**

For the cache to never have immutable entries that aren't immutable.

**`nix-env --version` output**

I don't know the Nix version that produced this invalid cache, but the faulty entry gets read with at least 2.13.7 to latest master (ef6d055ace0c3ccc1955ddf2150024ff9d58f2bd), didn't check other versions.

**Additional context**

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by infinisil** [comment]
This caused major curses in https://github.com/NixOS/nix.dev/pull/830#issuecomment-1922672937

**Comment by thufschmitt** [comment]
seems cursed indeed

@infinisil the nix.dev thread seems to point to a reliable way to reproduce this, but lost over five different messages. Can you summarize that in one easy-to-run example?

**Comment by infinisil** [comment]
@thufschmitt No that nix.dev thread is just a consequence of this. I don't have a way to reproduce this.

**Comment by ElvishJerricco** [comment]
Looks like I have some data in my fetcher cache indicative of this issue:

```
$ nix run nixpkgs#sqlite -- ~/.cache/nix/fetcher-cache-v1.sqlite "select * from cache where input like '%master%' and immutable = 1"
{"name":"source","type":"file","url":"https://github.com/thiagokokada/nix-alien/tarball/master"}|{"etag":"W/\"6e4793ed083206360adc8c2452919eac4cffe38c7219b7755f5caf72461db6cb\"","url":"https://codeload.github.com/thiagokokada/nix-alien/legacy.tar.gz/refs/heads/master"}|/nix/store/f7k9biy200vvq9nwqc1lgkjmf1n7s9ds-source|1|1706023784
{"name":"source","type":"file","url":"https://codeload.github.com/thiagokokada/nix-alien/legacy.tar.gz/refs/heads/master"}|{"etag":"W/\"6e4793ed083206360adc8c2452919eac4cffe38c7219b7755f5caf72461db6cb\"","url":"https://codeload.github.com/thiagokokada/nix-alien/legacy.tar.gz/refs/heads/master"}|/nix/store/f7k9biy200vvq9nwqc1lgkjmf1n7s9ds-source|1|1706023784
{"name":"source","type":"tarball","url":"https://github.com/thiagokokada/nix-alien/tarball/master"}|{"etag":"W/\"6e4793ed083206360adc8c2452919eac4cffe38c7219b7755f5caf72461db6cb\"","lastModified":1705306460}|/nix/store/7f9p6cx6a1z0msbd74bi14kj0qgarjmi-source|1|1706023784
```

I'm currently using Nix 2.18.1, and I'm *fairly sure* that's the version I was using when I last messed with that repo (the active NixOS generation during the date indicated in the timestamp in that output was using 2.18.1).

**Comment by fricklerhandwerk** [comment]
Related: https://github.com/NixOS/nix/issues/10146
